### PR TITLE
add Mistral-small-3.1 and Pixtral vision support

### PR DESCRIPTION
**Pull request by yicycyc**
## PR 新增模型支持：Mistral-small-3.1 Pixtral 

复现针对Mistral-small-3.1和Pixtral在transformers端的对应实现进行

mistral-small-3.1的视觉头复用了Pixtral，Pixtral在transformers中的实现只是一个视觉头，Pixtral-12b模型在transformers中通过llava式的方式调用pixtral视觉头和mistral文本组合得到，paddle缺乏前置，因此本PR实现的是Mistral-small-3.1的完整模型以及Pixtral的transformers实现。

精度验证均在缩层模型下进行。

## 主要改动

主要改动：

- 新增 `mistral3` 模型：
  - `Mistral3Config`
  - `Mistral3Model`
  - `Mistral3ForConditionalGeneration`
  - `Mistral3ForCausalLM` alias
  - HF safetensors / flex checkpoint AOA 转换规则
  - 支持图文前向与 SFT 训练

- 新增 `pixtral` 视觉塔：
  - `PixtralVisionConfig`
  - `PixtralVisionModel`
  - `PixtralImageProcessor`
  - `PixtralProcessor`

- 补充 Auto 映射：
  - `AutoConfig`
  - `AutoModel`
  - `AutoModelForCausalLM`
  - `AutoModelForConditionalGeneration`
  - `AutoProcessor`
  - `AutoImageProcessor`

- 补充多模态数据处理：
  - 支持 `image_sizes` 在 dataset/collate 流程中传递
  - 支持 Mistral3/Pixtral 图像 token 展开与图文输入构造
 

- 更新模型列表、能力矩阵与模型单测

## 前向对齐验证

模型：`mistral-small-3.1缩层模型`

两侧加载完全相同的 `.npy` 输入：

- `input_ids`
- `attention_mask`
- `pixel_values`
- `image_sizes`

输入样例为一张 `224x224` 图片加 prompt：

```text
Describe this image briefly.
```

总 token 数为 `256`。

### 结果

| 精度 | logits mean_diff | logits max_diff | 结论 |
| --- | ---: | ---: | --- |
| FP32 | `1.01e-06` | `1.72e-05` | 几乎零差异 |
| BF16 | `0` | `0` | 完全一致，Top-1 token 一致 |

## 生成对齐

Text-only 生成

Transformers 生成的 10 个 token:
[117577, 115201, 83673, 64162, 107744, 111937, 11254, 111937, 119792, 62615]
PaddleFormers 生成的 10 个 token:
[117577, 115201, 83673, 64162, 107744, 111937, 11254, 111937, 119792, 62615]

Multimodal 生成
输入：112x112 的测试图片 + "Describe this image." 文本 prompt

Transformers 生成的 10 个 token:
[64162, 18845, 14124, 16814, 5744, 31026, 33565, 34868, 14456, 61350]
PaddleFormers 生成的 10 个 token:
[64162, 18845, 14124, 16814, 5744, 31026, 33565, 34868, 14456, 61350]

两侧完全一致

### 1、文本

使用 GSM8K 做 BF16 full-SFT，Paddle 4 卡 sharding stage3 跑满 300 step，并与 Torch/ms-swift ZeRO-3 训练曲线对比。

共同设置：

- `max_seq_len = 512`
- `global batch size = 4`
- `max_steps = 300`
- `learning_rate = 1e-5`
- `warmup_steps = 0`
- `weight_decay = 0`
- `seed = 42`
- shuffle 关闭
- attention eager

训练结果：

| step | Paddle loss | Swift loss | diff |
| ---: | ---: | ---: | ---: |
| 1 | `13.359375` | `13.401804` | `-0.042429` |
| 2 | `9.484375` | `9.411279` | `0.073096` |
| 10 | `4.585938` | `4.784250` | `-0.198312` |
| 20 | `4.011719` | `4.166615` | `-0.154897` |
| 50 | `3.625000` | `3.814764` | `-0.189764` |
| 100 | `3.500000` | `2.874740` | `0.625260` |
| 150 | `3.425781` | `3.515725` | `-0.089944` |
| 200 | `3.250000` | `3.552930` | `-0.302930` |
| 243 | `3.207031` | `3.151531` | `0.055500` |
| 300 | `3.089844` | `3.604705` | `-0.514861` |

gsm8k_300_stage3_loss_curve

### 2、多模态

使用 https://github.com/PaddlePaddle/PaddleFormers/blob/develop/docs/zh/dataset_format.md#24-%E5%A4%9A%E6%A8%A1%E6%80%81%E6%8C%87%E4%BB%A4%E5%BE%AE%E8%B0%83sft%E6%95%B0%E6%8D%AE%E6%A0%BC%E5%BC%8F 数据集 做 BF16 full-SFT，Paddle 4 卡 sharding stage3 跑满 300 step，并与 Torch/ms-swift ZeRO-3 训练曲线对比。

- `max_seq_len / max_length = 4096`
- `global batch size = 4`
- `max_steps = 300`
- `learning_rate = 1e-5`
- `warmup_steps = 0`
- `weight_decay = 0`
- shuffle 关闭
- attention eager

训练结果：

| step | Paddle loss | Swift loss | diff |
| ---: | ---: | ---: | ---: |
| 1 | `12.687500` | `12.662024` | `0.025476` |
| 2 | `9.625000` | `9.594838` | `0.030162` |
| 3 | `9.625000` | `9.894209` | `-0.269209` |
| 10 | `7.515625` | `7.706814` | `-0.191189` |
| 20 | `6.234375` | `6.151044` | `0.083331` |
| 50 | `1.898438` | `1.336239` | `0.562199` |
| 100 | `0.507553` | `0.097169` | `0.410384` |
| 150 | `0.284668` | `0.120010` | `0.164658` |
| 200 | `0.155029` | `0.010435` | `0.144594` |
| 243 | `0.042799` | `0.630276` | `-0.587477` |
| 275 | `0.810303` | `0.016147` | `0.794156` |
| 300 | `0.578127` | `0.103580` | `0.474548` |

vl_300_loss_curve

### CODE REVIEW

File: paddleformers/transformers/pixtral/processor.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+import json
+import os
+
+from ..auto.tokenizer import AutoTokenizer
+from ..processing_utils import ProcessorMixin
+from .image_processor import PixtralImageProcessor
+
+
+class PixtralProcessor(ProcessorMixin):
```
Comment (risemeup1111) [inline paddleformers/transformers/pixtral/processor.py:28]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

`PixtralProcessor` 现在只有加载和 `model_input_names`，没有实现 Pixtral 必需的 `__call__` / `_get_num_multimodal_tokens` 展开逻辑。直接通过 `AutoProcessor.from_pretrained(...)(images=..., text=...)` 使用时会落到父类通用处理，文本里的一个 `[IMG]` 不会展开成 `H * W` 个 `[IMG]` 加 `[IMG_BREAK]` / `[IMG_END]`；随后 `Mistral3Model.get_placeholder_mask()` 只会数到 1 个 image token，而视觉塔会返回多 个 image features，首个多模态前向就会报 `Image features and image tokens do not match`。

请补齐 processor 侧的 Pixtral 输入构造，至少包含按 `image_sizes` 和 `patch_size * spatial_merge_size` 展开文本，并返回 tokenizer 输出与图像张量的合并结果。实现形状可以按下面收敛：

```python
def __call__(self, images=None, text=None, **kwargs):
    image_inputs = self.image_processor(
        images,
        patch_size=self.patch_size * self.spatial_merge_size,
        **image_kwargs,
    )

    expanded_text = []
    for prompt, (height, width) in zip(text, image_inputs["image_sizes"]):
        rows = int(height) // (self.patch_size * self.spatial_merge_size)
        cols = int(width) // (self.patch_size * self.spatial_merge_size)
        tokens = ([self.image_token] * cols + [self.image_break_token]) * rows
        tokens[-1] = self.image_end_token
        expanded_text.append(prompt.replace(self.image_token, "".join(tokens), 1))

    text_inputs = self.tokenizer(expanded_text, **text_kwargs)
    self._check_special_mm_tokens(expanded_text, text_inputs, modalities=["image"])
    return BatchFeature(data={**text_inputs, **image_inputs}, tensor_type=return_tensors)
```

File: paddleformers/transformers/pixtral/processor.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            tokenizer.convert_tokens_to_ids(image_break_token) if tokenizer is not None else None
+        )
+        self.image_end_token_id = tokenizer.convert_tokens_to_ids(image_end_token) if tokenizer is not None else None
+        self.image_max_pixels = image_max_pixels
+        self.image_min_pixels = image_min_pixels
+        self.video_max_pixels = video_max_pixels
+        self.video_min_pixels = video_min_pixels
+
+    @classmethod
+    def from_pretrained(cls, pretrained_model_name_or_path, **kwargs):
```
Comment (risemeup1111) [inline paddleformers/transformers/pixtral/processor.py]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这里覆盖了 `ProcessorMixin.from_pretrained`，但实现只按本地文件系统读 `preprocessor_config.json` / `processor_config.json` / `chat_template.json`，并且 `AutoTokenizer.from_pretrained` 没有透传 `cache_dir`、`revision`、`subfolder`、`download_hub` 等参数。对 Hub / ModelScope / AI Studio 的模型 ID，这些配置不会经过已有的 `resolve_file_path` 解析，`PixtralProcessor` 会退回到 `patch_size=16`、`spatial_merge_size=1` 等默认值，导致 Mistral3/Pixtral 的图像 token 展开数量和模型侧特征数量不匹配。

请删除这个覆盖方法，或改回走基类已经实现的加载路径；该路径会读取 `processor_config.json`、`preprocessor_config.json` 以及新的 `chat_template.jinja`，并透传下载参数。例如：

```python
@classmethod
def from_pretrained(cls, pretrained_model_name_or_path, **kwargs):
    return super().from_pretrained(pretrained_model_name_or_path, **kwargs)
```

如果确实需要保留自定义加载，也请复用 `get_processor_dict` / `_get_arguments_from_pretrained` / `from_args_and_dict`，不要直接用 `os.path.exists(os.path.join(...))` 读远程模型配置。

File: paddleformers/transformers/auto/modeling.py [hunk 3]
```diff
@@ -73,6 +73,8 @@
         ("Qwen3VLMoe", "qwen3_vl_moe"),
         ("Qwen3_5Moe", "qwen3_5"),
         ("Qwen3_5", "qwen3_5"),
+        ("Mistral3", "mistral3"),
+        ("Pixtral", "pixtral"),
```
Comment (risemeup1111) [inline paddleformers/transformers/auto/modeling.py:79]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

把 `Pixtral` 加到这里后，`AutoModel.from_pretrained` 对保存后的 `PixtralVisionModel` checkpoint 会从 `config.architectures=["PixtralVisionModel"]` 推断出 `model_name="PixtralModel"`，再尝试导入 `paddleformers.transformers.pixtral.modeling.PixtralModel`。当前模块只定义了 `PixtralVisionModel`，因此 PR 声明的 Pixtral `AutoModel` 映射在 vision-only 模型的保存/加载回环上会失败。

请给 `AutoModel` 的类名推断补特殊映射，或者提供兼容 alias，并补一条 `PixtralVisionModel.save_pretrained` 后 `AutoModel.from_pretrained` 的回归测试。可选修法之一：

```python
MAPPING_SPACIAL_KEY = OrderedDict(
    [
        ("Pixtral", "PixtralVision"),
        ("Gemma3", "Gemma3Text"),
        ("Ernie4_5_VLMoe", "Ernie4_5_VLMoeForConditionalGeneration"),
    ]
)
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已完成本轮审阅，发现几处需要先修复的兼容性问题，细节已放在行级评论里。主要集中在 Pixtral processor 的输入展开/加载路径，以及 Pixtral vision-only 模型的 AutoModel 映射闭环；这些问题会影响公开 API 的实际可用性。

CI 还有任务在运行或排队，修复后建议等待完整 CI 结果。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: tests/transformers/mistral3/test_modeling.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+# limitations under the License.
+from __future__ import annotations
+
+import tempfile
+import unittest
+
+import paddle
+from paddle import nn
+
+from paddleformers.cli.utils import get_lora_target_modules, get_multimodel_lora_target_modules
```
Comment (risemeup1111) [inline tests/transformers/mistral3/test_modeling.py]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

当前 Lint 失败在 isort：日志显示 `files were modified by this hook`，并且修复目标就是 `tests/transformers/mistral3/test_modeling.py`。这里新增的长导入没有按项目 isort 配置格式化，导致 CI 直接失败。请按 isort 输出格式调整后重新提交。

```suggestion
from paddleformers.cli.utils import (
    get_lora_target_modules,
    get_multimodel_lora_target_modules,
)
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交。此前的 Pixtral processor / AutoModel 映射相关 P1 问题尚未在本次提交中修改，仍需通过新的 commit 修复；另外当前 Lint 失败对应的新问题已放在行级评论里。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: tests/transformers/mistral3/test_modeling.py (hunk 4 above)
Comment (risemeup1111) [reply tests/transformers/mistral3/test_modeling.py]: 已复查当前提交，导入已按 isort 格式拆分，当前 Lint 也已通过。这条问题已经修复。

**Review by risemeup1111 — APPROVED** [review]
已复查当前提交，前面提到的 Pixtral processor / AutoModel 兼容性问题都已经收敛，当前没有看到新的阻塞点。`unittest-gpu-ci`、集成测试和 Lint 都已通过；`upload-coverage` 失败看起来是 runner 的 GLIBC 兼容问题，与本次改动无关。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/pixtral/processor.py (hunk 1 above)
Comment (yicycyc) [reply paddleformers/transformers/pixtral/processor.py:28]: 已在 6f7a7c20 中补齐 `PixtralProcessor.__call__` 与 `_get_num_multimodal_tokens`，按 `image_sizes` 和 `patch_size * spatial_merge_size` 展开 `[IMG]`，并新增 processor 回归测试。

File: paddleformers/transformers/pixtral/processor.py (hunk 2 above)
Comment (yicycyc) [reply paddleformers/transformers/pixtral/processor.py]: 已在 6f7a7c20 中删除本地文件专用的 `from_pretrained` override，恢复 `ProcessorMixin` 标准加载链路；新增测试覆盖 hub 参数、subfolder 与 download_hub 的透传。

File: paddleformers/transformers/auto/modeling.py (hunk 3 above)
Comment (yicycyc) [reply paddleformers/transformers/auto/modeling.py:79]: 已在 6f7a7c20 中提供 `PixtralModel = PixtralVisionModel` 兼容映射，并新增 save_pretrained 后通过 AutoModel 回载的回归测试。

**Comment by risemeup1111** [comment]
## Failed CI看板

| 流水线名称 | 问题标签 | 修复建议 | 日志片段 |
|------------|----------|----------|----------|
| [33036105321](https://github.com/PaddlePaddle/PaddleFormers/actions/runs/33036105321) | 未解析 | 无法进行 CI 根因分析：当前 PR 工作树中的 `agent/ci_analyzer` 目录缺失或不包含任何 Markdown 规则文件。  根据仓库强制规则，不能继续读取 CI 日志、判定根因或发布修复建议。 | 未提供 |

### 分析详情

无法进行 CI 根因分析：当前 PR 工作树中的 `agent/ci_analyzer` 目录缺失或不包含任何 Markdown 规则文件。

根据仓库强制规则，不能继续读取 CI 日志、判定根因或发布修复建议。

File: paddleformers/transformers/pixtral/modeling.py [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Paddle Pixtral Vision model."""
+
+import paddle
+from paddle import nn
+from paddle.distributed.fleet.utils import recompute
+
+from ...nn.norm import Norm as GeneralNorm
```
[the remark this answers is not in the document]
Comment (yicycyc) [reply paddleformers/transformers/pixtral/modeling.py:21]: 确认该问题存在，已在 `c20b440` 修复：Mistral3 不再导入已删除的 `PixtralRMSNorm`，projector 改用 `GeneralNorm.create`，并保留原 hidden size 与 epsilon。现有 Mistral3 建模测试在模块收集阶段也会覆盖该导入回归。
